flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -347,7 +348,13 @@ private void setChaining(Map<Integer, byte[]> hashes, List<Map<Integer, byte[]>>
         final Map<Integer, OperatorChainInfo> chainEntryPoints =
                 buildChainedInputsAndGetHeadInputs(hashes, legacyHashes);
         final Collection<OperatorChainInfo> initialEntryPoints =
-                new ArrayList<>(chainEntryPoints.values());
+                chainEntryPoints.values().stream()
+                        .sorted(
+                                Comparator.comparing(
+                                        operatorChainInfo ->
+                                                hashes.get(operatorChainInfo.getStartNodeId()),
+                                        StreamingJobGraphGenerator::compareHashes))
```

Comment:
Is it actually possible to have multiple sources with the same startNodeId?
If so, I think hash comparison is not covered by tests (please feel free to leave it as is if it would require too much effort).

Reply:
No, `nodeId` is incremented in the `StreamEnvironment` for each added node. So each job vertex has a unique id, which may change on restart (that's why the hashes are used). `startNodeId` is the `nodeId` for the particular chain.

 There are also a couple of places where the nodes are indexed by the `nodeId`, so duplicates would break much more.

Reply:
Then shouldn't hashes be compared first (before nodeId)?

Reply:
The ids are not regenerated on failover restart but on manual restart (as in UCRescaleITCase). The ids are actually good enough to guarantee stability and I switched form hashes to id since a couple of tests rely on the declaration order of sources.